firedrake/mesh.py

```diff
@@ -2457,51 +2458,58 @@ def clear_spatial_index(self):
         the coordinate field)."""
         self._spatial_index = None
 
-    @property
-    def spatial_index(self):
-        """Spatial index to quickly find which cell contains a given point.
-
-        Notes
-        -----
-
-        If this mesh has a :attr:`tolerance` property, which
-        should be a float, this tolerance is added to the extrama of the
-        spatial index so that points just outside the mesh, within tolerance,
-        can be found.
+    @utils.cached_property
+    def bounding_box_coords(self):
+        """Calculates bounding boxes for spatial indexing.
 
+        If we have a higher-order (bendy) mesh we project the mesh coordinates into
+        a Bernstein finite element space. Functions on a Bernstein element are
+        Bezier curves and are completely contained in the convex hull of the mesh nodes.
+        Hence the bounding box will contain the entire element.
         """
         from firedrake import function, functionspace
         from firedrake.parloops import par_loop, READ, MIN, MAX
 
-        if (
-            self._spatial_index
-            and self.coordinates.dat.dat_version == self._saved_coordinate_dat_version
-        ):
-            return self._spatial_index
-
         gdim = self.geometric_dimension()
         if gdim <= 1:
             info_red("libspatialindex does not support 1-dimension, falling back on brute force.")
             return None
 
+        coord_element = self.ufl_coordinate_element()
+        coord_degree = coord_element.degree()
+        if ufl.cell.simplex(self.topological_dimension()) != self.ufl_cell():
+            # Non-simplex element, e.g. quad or tensor product
+            mesh = self
+        elif coord_degree == 1:
+            mesh = self
+        elif coord_element.family() == "Bernstein":
+            # Already have Bernstein coordinates, no need to project
+            mesh = self
+        else:
+            # For bendy meshes we project the coordinate function onto Bernstein
+            bernstein_fs = functionspace.VectorFunctionSpace(self, "Bernstein", coord_degree)
+            f = function.Function(bernstein_fs)
+            f.project(self.coordinates)
```

Comment:
Bernstein supports `interpolate` by using linear combinations of the "spectral" dofs.  Since coordinates are polynomials, the interpolation should be lossless.

https://github.com/firedrakeproject/fiat/pull/124